examples/IMUKittiExampleGPS.cpp

```diff
@@ -0,0 +1,293 @@
+
+    // Set ISAM2 parameters and create ISAM2 solver object
+    ISAM2Params isamParams;
+    isamParams.factorization = ISAM2Params::CHOLESKY;
+    isamParams.relinearizeSkip = 10;
+
+    ISAM2 isam(isamParams);
+
+    // Create the factor graph and values object that will store new factors and values to add to the incremental graph
+    NonlinearFactorGraph newFactors;
+    Values newValues; // values storing the initial estimates of new nodes in the factor graph
+
+    /// Main loop:
+    /// (1) we read the measurements
+    /// (2) we create the corresponding factors in the graph
+    /// (3) we solve the graph to obtain and optimal estimate of robot trajectory
+    printf("-- Starting main loop: inference is performed at each time step, but we plot trajectory every 10 steps\n");
+    size_t imuMeasurementIndex = 0;
+    for (size_t gpsMeasurementIndex = firstGPSPose; gpsMeasurementIndex < GPS_measurements.size() - 1; gpsMeasurementIndex++) {
+        // At each non=IMU measurement we initialize a new node in the graph
+        auto currentPoseKey = X(gpsMeasurementIndex);
+        auto currentVelKey = V(gpsMeasurementIndex);
+        auto currentBiasKey = B(gpsMeasurementIndex);
+        double t = GPS_measurements[gpsMeasurementIndex].Time;
+
+        if (gpsMeasurementIndex == firstGPSPose) {
+            // Create initial estimate and prior on initial pose, velocity, and biases
+            newValues.insert(currentPoseKey, currentPoseGlobal);
+            newValues.insert(currentVelKey, currentVelocityGlobal);
+            newValues.insert(currentBiasKey, currentBias);
+            newFactors.add(PriorFactor<Pose3>(currentPoseKey, currentPoseGlobal, sigma_init_x));
+            newFactors.add(PriorFactor<Vector3>(currentVelKey, currentVelocityGlobal, sigma_init_v));
+            newFactors.add(PriorFactor<imuBias::ConstantBias>(currentBiasKey, currentBias, sigma_init_b));
+        } else {
+            double t_previous = GPS_measurements[gpsMeasurementIndex-1].Time;
+
+            // Summarize IMU data between the previous GPS measurement and now
+            currentSummarizedMeasurement = std::make_shared<PreintegratedImuMeasurements>(IMU_params, currentBias);
+            static size_t includedIMUmeasurementCount = 0;
+            while (imuMeasurementIndex < IMU_measurements.size() && IMU_measurements[imuMeasurementIndex].Time <= t) {
+                if (IMU_measurements[imuMeasurementIndex].Time >= t_previous) {
+                    currentSummarizedMeasurement->integrateMeasurement(IMU_measurements[imuMeasurementIndex].Accelerometer, IMU_measurements[imuMeasurementIndex].Gyroscope, IMU_measurements[imuMeasurementIndex].dt);
+                    includedIMUmeasurementCount++;
+                }
+                imuMeasurementIndex++;
+            }
+
+            // Create IMU factor
+            auto previousPoseKey = X(gpsMeasurementIndex-1);
```

Comment:
same comment about keys in-pace -> X(i-1)

Reply:
Addressed in b48bdfdc696328be3fedfe11db18b83d3341e7bc

Reply:
Addressed in b48bdfdc696328be3fedfe11db18b83d3341e7bc